### PR TITLE
[SDK-1895] Add options to match HTTP requests based on HTTP method

### PR DESCRIPTION
**Pull request by stevehobbsdev**
This PR adds an option for the HTTP Interceptor so that routes can be matched
based on HTTP method. This means that different scopes can be applied to
different routes depending on whether it is a GET or a POST (amongst other
verbs).

This PR also removes the custom `WindowService`, instead favouring use of the
build-in `Document` and `Location` types for the same functionality.

### CODE REVIEW

File: projects/playground/src/app/app.component.spec.ts [hunk 1]
```diff
@@ -127,7 +118,7 @@ describe('AppComponent', () => {
       expect(authMock.logout).toHaveBeenCalledWith({
         localOnly: false,
         federated: false,
-        returnTo: 'http://localhost:4200',
+        returnTo: 'http://localhost:9876',
```
Comment (lbalmaceda) [inline projects/playground/src/app/app.component.spec.ts:121]: I noticed this changed. Why is that? Does it make sense to try to fix it to the previous value?

File: projects/playground/src/app/app.component.ts [hunk 2]
```diff
@@ -53,8 +50,9 @@ export class AppComponent {
     const options: LogoutOptions = {
       localOnly: formOptions.localOnly === true,
       federated: formOptions.federated === true,
-      returnTo: this.window.location.origin,
+      returnTo: this.doc.location.origin,
```
Comment (lbalmaceda) [inline projects/playground/src/app/app.component.ts:53]: How does this "doc" behave vs the "window" object in terms of Angular SSR?

File: projects/auth0-angular/src/lib/auth.interceptor.ts [hunk 3]
```diff
@@ -113,6 +113,13 @@ export class AuthHttpInterceptor implements HttpInterceptor {
     };
 
     if (isHttpInterceptorRouteConfig(route)) {
+      if (
+        route.httpMethod &&
+        route.httpMethod.toLowerCase() !== request.method.toLowerCase()
```
Comment (lbalmaceda) [inline projects/auth0-angular/src/lib/auth.interceptor.ts]: I think we should be comparing against the upper case, without transforming if we want to fail when there's no match. According to this, it looks like method names are case sensitive https://tools.ietf.org/html/rfc7230#section-3.1.1 and https://tools.ietf.org/html/rfc7231#section-4. But we can sync and evaluate this, e.g. what other libraries do

**Review by lbalmaceda — CHANGES_REQUESTED** [review]

File: projects/playground/src/app/app.component.spec.ts (hunk 1 above)
Comment (stevehobbsdev) [reply projects/playground/src/app/app.component.spec.ts:121]: The actual value doesn't really matter, just that it matches what is provided by the window. Previously we were specifically mocking `WindowService`, but now it uses the built-in `Document` type which has a different value. It's not hugely important to the test, the key is that `returnTo` was specified.

File: projects/playground/src/app/app.component.ts (hunk 2 above)
Comment (stevehobbsdev) [reply projects/playground/src/app/app.component.ts:53]: It behaves better as I believe Angular itself can substitute this with something that works in SSR. However, it doesn't itself fix SSR support as there are other things within the underlying SDK that are preventing SSR from being supported fully.

File: projects/auth0-angular/src/lib/auth.interceptor.spec.ts [hunk 4]
```diff
@@ -23,17 +19,27 @@ describe('The Auth HTTP Interceptor', () => {
   let req: TestRequest;
   const testData: Data = { message: 'Hello, world' };
 
-  const assertAuthorizedApiCallTo = (url: string, done: any) => {
-    httpClient.get(url).subscribe(done);
+  const assertAuthorizedApiCallTo = (
+    url: string,
+    done: () => void,
+    method = 'get'
```
Comment (lbalmaceda) [inline projects/auth0-angular/src/lib/auth.interceptor.spec.ts:25]: I was expecting tests with lowercase method names to fail 🤔

File: projects/auth0-angular/src/lib/auth.config.ts [hunk 5]
```diff
@@ -1,6 +1,18 @@
 import { CacheLocation, GetTokenSilentlyOptions } from '@auth0/auth0-spa-js';
 import { InjectionToken } from '@angular/core';
 
+/**
+ * Defines a common set of HTTP methods.
+ */
+export const enum HttpMethod {
+  Get = 'GET',
```
Comment (lbalmaceda) [inline projects/auth0-angular/src/lib/auth.config.ts:8]: I'm not 110% convinced about the casing.

The examples I found are either
1. method names using lower. e.g.: `httpClient.post("your URL here")`
2. param values using uppercase. e.g.: `httpClient.request("POST", "your URL here")`

However, this [SO post](https://stackoverflow.com/questions/3069743/coding-conventions-naming-enums) talks about different conventions, so I ended up more confused than I was 😂

File: projects/auth0-angular/src/lib/auth.config.ts [hunk 6]
```diff
@@ -52,6 +64,15 @@ export interface HttpInterceptorRouteConfig {
...
   tokenOptions?: GetTokenSilentlyOptions;
+
+  /**
+   * The HTTP method to match on. If specified, the HTTP method of
+   * the outgoing request will be checked against this. If there is no match, the
+   * Authorization header is not attached.
+   *
+   * The HTTP method name is case-sensitive.
+   */
+  httpMethod?: HttpMethod | string;
```
Comment (lbalmaceda) [inline projects/auth0-angular/src/lib/auth.config.ts:75]: do we want to keep this `string` out of the API for now, until someone requires to use a method we don't have in the enum? Thinking removing it later will be a breaking change, and that supported methods should be those that the angular HTTP client supports.

```suggestion
  httpMethod?: HttpMethod;
```

**Review by lbalmaceda — CHANGES_REQUESTED** [review]

File: projects/auth0-angular/src/lib/auth.interceptor.spec.ts (hunk 4 above)
Comment (stevehobbsdev) [reply projects/auth0-angular/src/lib/auth.interceptor.spec.ts:25]: `method` here refers to which method on `HttpClient` to call, the names of which follow normal camelCase convention.

File: projects/auth0-angular/src/lib/auth.config.ts (hunk 5 above)
Comment (stevehobbsdev) [reply projects/auth0-angular/src/lib/auth.config.ts:8]: I think you're confusing the method names of `HttpClient` vs. HTTP verbs.

The methods on `HttpClient` (`get`, `post`, `delete`, etc) follow normal JS camelCase convention. The HTTP verbs/methods themselves (`GET`, `POST`, `DELETE`) appear to be uppercase as we previously discussed and are case sensitive.

There's a third convention at play in this snippet, which is TypeScript enums. These appear in TitleCase in [the TypeScript docs](https://www.typescriptlang.org/docs/handbook/enums.html), so I've gone with that convention here.

File: projects/auth0-angular/src/lib/auth.config.ts (hunk 6 above)
Comment (stevehobbsdev) [reply projects/auth0-angular/src/lib/auth.config.ts:75]: I don't have a particularly strong opinion on this, but I will say that Angular does let you pass _any_ verb as string [using `request`](https://www.typescriptlang.org/docs/handbook/enums.html). If we removed support for string, we'd already be limiting customers needlessly if they need that option.

File: projects/auth0-angular/src/lib/auth.interceptor.spec.ts (hunk 4 above)
Comment (lbalmaceda) [reply projects/auth0-angular/src/lib/auth.interceptor.spec.ts:25]: right this is asserting the mocked call, not the configuration the developer has to provide.

File: projects/auth0-angular/src/lib/auth.config.ts (hunk 5 above)
Comment (lbalmaceda) [reply projects/auth0-angular/src/lib/auth.config.ts:8]: Sounds good. Thanks for explaining that.

File: projects/auth0-angular/src/lib/auth.config.ts (hunk 6 above)
Comment (lbalmaceda) [reply projects/auth0-angular/src/lib/auth.config.ts:75]: That's a fair point.

**Review by lbalmaceda — APPROVED** [review]

**Pull request merged**